MoreLinq/Duplicates.cs

```diff
@@ -0,0 +1,61 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2023 Julien Aspirot. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        ///   Returns all duplicated elements of the given source.
+        /// </summary>
+        /// <param name="source">source sequence.</param>
+        /// <typeparam name="T">The type of the elements in the source sequence.</typeparam>
+        /// <returns>all elements that are duplicated.</returns>
+        public static IEnumerable<T> Duplicates<T>(this IEnumerable<T> source) => Duplicates(source, null);
+
+        /// <summary>
+        ///   Returns all duplicated elements of the given source, using the specified element equality comparer
+        /// </summary>
+        /// <param name="source">source sequence.</param>
+        /// <param name="comparer">The equality comparer to use to determine whether or not keys are equal.
+        /// If null, the default equality comparer for <c>TSource</c> is used.</param>
+        /// <typeparam name="TSource">Type of the source sequence</typeparam>
+        /// <returns>all elements of the source sequence that are duplicated, based on the provided equality comparer</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
+        public static IEnumerable<TSource> Duplicates<TSource>(this IEnumerable<TSource> source, IEqualityComparer<TSource>? comparer)
+        {
+            if (source is null) throw new ArgumentNullException(nameof(source));
+
+            return GetDuplicates();
+
+            IEnumerable<TSource> GetDuplicates()
+            {
+                var keySet = new HashSet<TSource>(comparer);
+                var yieldSet = new HashSet<TSource>(comparer);
```

Comment:
would probably better as:
```cs
var counts = new Dictionary<TSource, int>(comparer);
...

if (!counts.TryGetValue(element, out var count))
  count = 0;
if (count == 1)
  yield return element;
counts[element] = count + 1;

Reply:
@viceroypenguin i tested it out with .NET Benchmark using a list of int of 15k elements for i from 1 to 15k, with duplicates for numbers that  are multiples of 100.

![image](https://github.com/morelinq/MoreLINQ/assets/14810960/a0ce267f-6966-4c78-bbff-c12e882f9f00)

it looks like the Hashset approach is 
1- faster, i would have expected similar performance  
2- consume less memory ( that i am a bit surprised, since we have 2 data structure in memory )

but i did not test with a custom Equality comparer

Reply:
A majority of the cost here is going to be in memory allocation; so doubling the memory requirements (double <TSource> sets) doubles time cost as well.